### PR TITLE
Fix #87: Fix plane create_from_position / point_height_above_plane

### LINKED ISSUE #87: Inconsitent Position / Distance to plane
Thank you for this useful Library.
There is some unexpected behavior for planes concerning the sign of the distance:
Take for example the following Plane:
```
plane = pyrr.plane.create_from_position(position=[0, 0, 1], normal=[0, 0, 1])
```
which returns the expected plane: `array([ 0,  0,  1, -1])`. However `pyrr.plane.position(plane)` returns `array([ 0,  0, -1])`, instead of `[0, 0, 1]`
 A related issue can be observed when calculating the distance of a point calculated to lie on the plane, which should therefore be 0:
```pyrr.geometric_tests.point_height_above_plane(pyrr.geometric_tests.point_closest_point_on_plane([0, 0, 1], plane), plane)```
but returns -2

### PR DESCRIPTION
**Pull request by adamlwgriffiths**
Fix plane distance being calculated as inverse in
point.create_from_position which caused an error in
create.
Fix geometric_tests.point_height_above_plane
calculating incorrect values.

### CODE REVIEW

**Comment by refelix** [comment]
I think that fix could make the library consistent, but not the standard way.
In http://mathworld.wolfram.com/HessianNormalForm.html it states 

> Here, the sign of `p` determines the side of the plane on which the origin is located. If `p>0`, it is in the half-space determined by the direction of `n^`, and if `p<0` , it is in the other half-space. 

However your new definition of `create_from_position()`  the sign is not correct in my opinion.

So instead of changing the sign in height_above_plane and create_from position, I would rather prefer to change the sign in `plane.position()` and `geometric_tests.point_closest_point_on_plane()`

I guess `ray_intersect_plane`  will also be affected by the definition of the sign of `d` (or `p` respectively). I can look into it once you communicated if you want to use the standard definition or if you want to use the definition used in this fix for some reason.

**Comment by adamlwgriffiths** [comment]
Thank you for finding that, it's been a long time since I worked on that code and I was wracking my brain trying to figure out why d was -ve. I'll revert that change and fix the other functions that misuse that value.

**Comment by refelix** [comment]
That sounds good. Thanks for looking into the issue so fast.

**Comment by adamlwgriffiths** [comment]
Ok, so I've pushed a fix that reverts the -d/+d change and fixes the geometry tests.
I am however, unhappy about some of these functions, at least in terms of naming.
Distance being negative is unexpected, even with the comment it's error-prone.
I'm up for suggestions as to nomenclature that can help reduce user error.

**Comment by adamlwgriffiths** [comment]
Ok so looking into it further, I think this is confusion between terminology.
The create functions are relative to the origin, not relative to a point in space.
Hence, my use of distance and not "p".
The latest commit moves away from this definition and toward the "plane relative to point" point of view, which if far less intuitive.
I believe I had it correct with the first commit.

**Comment by adamlwgriffiths** [comment]
I will say if you believe you have a better suggestion, I'm all ears.
I don't really have time to look into these matters properly.

**Comment by refelix** [comment]
The confusion may arise from the fact that the distance stays the same, whether you say "distance from point to plane" or "distance from plane to point", even the sign of the connecting vector changes. However the sign of the distance is only defined by the normal of the plane and on which side the point is. So actually your `p` is the [(signed) Distance](http://mathworld.wolfram.com/Point-PlaneDistance.html) of the origin from ( or to) the plane. If you insert `x0 = [0, 0, 0]` into (13) of the  link, you obtain `D = n*(x0-x1) = - n*x1 = p`

It also took me a while to get a clear comprehension of the sign, but I hope I could convince you that your original create functions followed the canonical representation of the Hessian normal form.

There is still a mistake in the `point_closest_point_on_plane()` function maybe related to the mix up of directions:
```
p = np.array([0,0,1])
n = np.array([0,0,1])
pl = pyrr.plane.create_from_position(p,n)
pyrr.geometric_tests.point_closest_point_on_plane(p, pl)
```
returns `array([ 0,  0, -1])` 
you could use 
```
point - n * (point - pyrr. plane.position(pl))
```
you can  fix this with 
changing `d = np.dot(p, n)` to `d = np.dot(-p, n)` but to avoid the confusion with all the `p` and `d` with the` plane[3]`, I suggest skipping the definition of p and using  `d = np.dot(pyrr.plane.position(pl), n)`

By the way I liked your effort of the first commit to replace `pl[3]` with `pyrr.plane.distance(pl)`

I hope I could fix some confusion.

File: pyrr/geometric_tests.py [hunk 1]
```diff
@@ -316,17 +316,17 @@ def point_height_above_plane(point, plane):
         negative if the point is behind the plane.
     """
     """
-    http://www.vitutor.com/geometry/distance/point_plane.html
-    d(P) = (AX + BY + CZ + D) / sqrt(A^2 + B^2 + C^2)
-
-    Normal is unit length, so it's length is 1.0.
-    Therefore, we can ignore the division all together.
-    Just perform Pn . [XYZ1]
+    Because we store normalised normal, we can simply
```
Comment (refelix) [inline pyrr/geometric_tests.py]: In my oppinion this is a good Idea, but I think the create Functions should make sure that this is the case

Comment (adamlwgriffiths) [reply pyrr/geometric_tests.py]: It should really have a normalise call in there, there's no guarantee in the type that its normalised. You could pass in your own hand-made data.

**Comment by adamlwgriffiths** [comment]
Would you be able to fork and demonstrate the changes your suggesting?
I'm still not 100% if you're suggesting that a plane along x,y at z=1 facing along +z (p = [0,0,1], n = [0,0,1]) would have a distance of +1 or -1 for the creation functions.
I would assume +1 as it is the most intuitive (relative to origin). If not, it seems we have a conflict between the mathematical representation and the one we are using here.

**Comment by refelix** [comment]
I created a pull request where I fixed the `point_closest_point_on_plane()`, and also added a normalize in the create function. 

With regard to the intuitiveness of the distance: If it is just about the `create_xy()` and related functions: I agree that it is charming to say I want a plane at distance `z` from the origin, and obtain a plane going through `z`. You could git this by setting invert = True by default. or, instead of calling it distance, you provide `intersect = -distance` as an argument to the function. If you do so, you should invert not only the normal but the created plane in order to to obtain the intersection when `invert == True`.

File: pyrr/geometric_tests.py (hunk 1 above)
Comment (refelix) [reply pyrr/geometric_tests.py]: If you think people will use non normalized planes, then `plane.normal` and `plane.distance` should be divided by `np.linalg.norm(pl[:3])`. Personally I would be fine in saying the function wants a plane that is in accordance to the Hessian normal form

**Comment by adamlwgriffiths** [comment]
Apologies, I haven't had much time lately to look at this.

I think we should do one of the following:
1. Correct the changes made in this PR so that +ve distance moves the plane along the normal. Ie [0,0,1,1] will be a plane at z=1 with normal of [0,0,1].
2. Re-write the plane module to use a different representation based on the hessian normal form representation.
3. Both of the above.

We should definitely do 1.
To use a metaphor, a car travelling along it's direction vector increases its overall distance travelled, not decreases. The plane definition, being relative to the origin, should act in a similar way.

Doing 2 or 3 would break the existing API, but I'm happy to do that and bump the major version.

**Comment by refelix** [comment]
Commenting on your metaphor: The point is, it is really arbitrary if you use the distance traveled or the distance to go. Let me provide an example showing the sign of the distance as chosen in the Hessian normal form is perfecly resonable:
Imagine the distance from your home to work is 1km (equal to the distance from work to home).
Now  you start driving, at some Point work is still 900m to go while you left your home 100m behind you. If you have to give one of the distances the negative sign, I think choosing the Point behind you as negative distance is perfectly reasonable. 
A related situation where you have signed distances would be if you take the distance to the surface of a cylinder: defining outside as positive distance and inside negative allows a continuous transition to a line when you reduce the diameter of the cylinder. If the normal of the cylinder points outwards, you have the same definition as for the plane, positive distance when the normal is pointing in the direction of the point. Of course my argument fails if you say by default the normal of the cylinder has to point inward, so my strongest argument for using the hessian normal form representation is that it makes the exchange with other libraries easier (for example pointcloud library), and it is easier to implement formulas. 

You make it sound like a bigger effort to go along with the hessian normal form, I'd argue currently the complete plane module is broken, at parts using formulas dependent on the hessian normal form, at part using the other sign of the distance. 
So making it consistent with the references you provide would be to most straight forward thing to do in my opinion and does not even require a rewrite of the api.

**Comment by adamlwgriffiths** [comment]
Apologies for the delay. I've pushed an update that reverts distance to be along the normal.
I also found an issue with some of the create_{xy,xz,yz} functions not inverting properly.

My concern is just about representing the data in the most appropriate format.
If the current format isn't appropriate then perhaps we should consider ancestoring it and creating a different version.
The current version will function regardless. It's just nice to be consistent with the rest of the world =P.

Let me know if there's any problems with the code.

**Comment by ywy1217** [comment]
seeing the comment    # -d = a * x  + b * y + c * z, but then you create(n, -d)

I think the confusion really starts from the definition of plane. It would be nice that you point it out at the very beginning that your (4,) array (A,B,C,D) is Ax+By+Cz-D=0, instead of the definition ax+by+cz+d=0, or P .* n + d = 0, where P=(x,y,z), n=(a, b, c), in the wiki link you posted: http://mathworld.wolfram.com/Plane.html
You would also want to correct the description at the beginning of pyrr.plane:
----
The fourth value is the distance of the origin from the plane, down the normal.
A negative value indicates the origin is behind the plane, relative to the normal.
----
It is contradicting your comment inside the relevant functions. You really mean that it is the distance of plane to the origin, not the flipped. There might be other places the two definitions are conflicting against each other, either in the comment or in the code.

There is another issue in a relevant function pyrr.geometric_tests.ray_intersect_plane, where it seems to treat the ray as line by allowing t<0 solution.

**Pull request merged**